Updated q_learning doc and added seed to blackjack notebook

acyu9 opened the pull request:
This is a draft pull request to open discussion about a better way of setting seed for the Blackjack Jupyter notebook example. 

The `blackjack.ipynb` cell 2 currently uses `seed.py` to call `set_seed`. However, when running cell 3 multiple times, it was not producing the same results (mean test scores). When adding `base_env.reset()` and `np.random.seed()` in cell 3, I was able to get consistent results. Before making any changes, I wanted to confirm if `seed.py` is intended to handle setting the seed for algorithms as well, since I’ve only seen it used in `compare_algorithms.py`. If `seed.py` isn't meant for algorithm seeding, I can add a new function in seed.py to handle `base_env.reset()` and `np.random.seed()` to keep all seed-related functionalities together, and then add that to `blackjack.ipynb` as an example for the students.

Other minor updates:
- ` blackjack.ipynb’s q_learning` in cell 3 now returns `rewards` to match the newly added return variable.
- For `rl.py`, I updated the `q_learning` docstring: the descriptions for `pi` and `V` were swapped to match the return order (`Q, V, pi, Q_track, pi_track, rewards` instead of `Q, pi, V`), and I added a description for `rewards`.

I would also like to propose the following documentation updates, as I found some areas confusing during this project:
- In `plots.ipynb`, briefly explain the Blackjack policy heatmap (e.g., what the x- and y-axes represent).
- In `README.md`, categorize the available environments (e.g., grid world: Frozen Lake, Taxi; non-grid world: Blackjack, Cartpole, Acrobat, Pendulum) and briefly explain the purpose of wrappers (e.g., wrappers like Cartpole allow for state space size adjustments).

I’m fairly new to this process, so I apologize if this is not following the proper procedure. Your feedback is greatly appreciated. Thank you!

ethankoch4 commented:
Thanks for the discussion opener. Yes, this is a `gymnasium` implementation detail that's a bit unfortunate as it makes for a confusing user experience.

I don't super love it, but likely the best way to do this for now is to call `env.reset` in the init methods and possibly in the test_env methods when it's passed in.

ethankoch4 commented on bettermdptools/algorithms/rl.py:
```diff
@@ -141,17 +141,20 @@ def q_learning(self,
             Log of complete policy for each episode
+        
+        rewards {numpy array}, shape(n_episodes):
+            Total reward obtained in each episode.
```
thanks for catching this

ethankoch4 commented:
Please pull in the latest changes and feel free to add as much documentation as you can!

As for the seed, we also welcome contributions on that :)

acyu9 commented:
Changes made: 
- Added seed in TestEnv.test_env()
- Added seed in RL init
- Added seed sample usage in blackjack notebook
- Added brief explanation on how to interpret blackjack policy heat map in plots notebook
- Categorized available wrappers as grid/non-grid worlds in readme

ethankoch4 commented on README.md:
```diff
@@ -21,6 +21,22 @@ pip install bettermdptools
+
+`bettermdptools` provides wrappers for both **grid-world** and **non-grid-world** environments. These wrappers enhance the functionality of the original environments, such as discretizing continuous state spaces. Some of these wrappers also allow adjustment of the state space size, such as in Cartpole.
+
+See the [`bettermdptools/bettermdptools/envs`](bettermdptools/bettermdptools/envs) directory for more details.
```
Could you actually add a link to the documentation here? And add more description to the documentation if it isn't currently sufficient

ethankoch4 commented on bettermdptools/algorithms/rl.py:
```diff
@@ -31,8 +31,9 @@
 
 class RL:
-    def __init__(self, env):
+    def __init__(self, env, seed=None):
```
Actually I think we should just import the seed from seed.py so users don't have to keep passing it around in multiple places

ethankoch4 reviewed (changes requested):
Thanks so much for your PR!  A few changes, but then I think it will be perfectly ready to merge :)

acyu9 commented:
I made the suggested changes:
- pdoc is regenerated to include the updates
- user can now call set_seed() to automatically set the seed for test_env and RL init. An example of this is available in the Blackjack notebook.
Thank you!

ethankoch4 commented on bettermdptools/algorithms/rl.py:
```diff
@@ -44,6 +45,10 @@ def __init__(self, env):
         #       return np.random.randint(len(Q[state]))
+        seed = get_seed()
+        if seed is not None:
+            np.random.seed(seed)
```
```suggestion
```

This is already done in the seed.py

acyu9 replied:
I included it because I noticed the results are different for q learning.
Without calling `np.random.seed(seed)` again: `-0.11308904 mean Q` and `-0.11 mean test_scores`
Calling `np.random.seed(seed)` again: `-0.1253272 mean Q` and `-0.03 mean test_scores`
I thought it would be better to call it again to ensure all the possible randomness are properly seeded but I can remove it if that's not needed.

ethankoch4 commented on bettermdptools/algorithms/rl.py:
```diff
@@ -44,6 +45,10 @@ def __init__(self, env):
+        seed = get_seed()
+        if seed is not None:
+            np.random.seed(seed)
+            env.reset(seed=seed)
```
I think this should actually be put in the initialization of each of the environment wrappers right? and _also_ wherever `env.reset` is currently called.

acyu9 replied:
I put the seed functionality in `test_env()` and `RL()` because if I initialize the wrapper with seed then run VI, PI, and QL, the environment only resets for VI (mean test_scores are `-0.13 , -0.01, -0.12`, respectively). Putting the seed in `test_env()` resets the env every time each algorithm is called (mean test_scores are `-0.13, -0.13, and -0.1`). 

Regarding your suggestion to set the seed in `env.reset()`, are you suggesting that every time `env.reset()` is called, the seed should automatically be set based on the global seed, like `env.reset(seed=SEED)`? If so, could you provide some guidance on how I might go about ensuring this?

As of my understanding, I only call `env.reset()` once before running each algorithm, and this is already handled with the seed implementation I added in `test_env()` and `RL()`. In the current code, the only place I see `env.reset()` being explicitly called is in the q_learning function: `state, info = self.env.reset()`. However, this doesn’t seem to directly allow for setting the seed in this context.

Thank you!
